### PR TITLE
[SPARK-42038][SQL] SPJ: Support partially clustered distribution

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

Currently with [storage-partitioned join](https://docs.google.com/document/d/1foTkDSM91VxKgkEcBMsuAvEjNybjja-uHk-r3vtXWFE/edit#heading=h.82w8qxfl2uwl), both sides of the join must be **fully clustered** over the partition values, that is, each Spark partition should have a distinct partition value. To guarantee this, Spark will group all the input partitions reported by a V2 data source on the partition values. The consequence, however, is that it can easily lead to data skew, when a few partitions have large amount of data.

This PR introduce a new mechanism which requires only one side of the storage-partitioned join to be fully clustered, while the other side can be **partially clustered**, i.e., there could exist multiple Spark partitions with the identical partition value. At planning time, Spark will compare the statistics from both sides of the join, and pick the side with smaller size as full clustered side, while the other side is partially clustered. It then replicate the partitions on the former to match the number of partitions on the latter. 

The concept of this optimization is similar to the existing `OptimizeSkewedJoin`, and techniques such as key salting, but is applied on DataSource V2 level instead, and doesn't require AQE to be enabled. Unlike `OptimizeSkewedJoin`, however, this optimization is applied before any shuffle happens.

A new config `spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled` is added to enable & disable the feature. By default it is turned off.

For instance, consider the following SQL query:
```sql
SELECT * FROM a JOIN b on a.id = b.id
```

while table `a` reports partitions `[0, 1, 2]` while `b` reports partitions `[0, 1, 1, 1, 2, 2]`.

Without the PR, Spark would group input partitions so that both sides will have 3 partitions `[0, 1, 2]`. With the PR, Spark will choose the right-hand side as partially clustered and match the left-hand side with it. Therefore, both sides have 6 partitions `[0, 1, 1, 1, 2, 2]`. 

Note this PR currently relies on a simple heuristic and always pick the side with less data size based on table statistics as the side fully clustered, even though it could also contain skewed partitions. In future, we can potentially do fine-grained comparison based on partition values.

### Why are the changes needed?

As mentioned in the previous section, this feature can help to reduce data skewness during storage-partitioned join, when a few partitions are mapped to a large amount of data.

### Does this PR introduce _any_ user-facing change?

A new Spark config, `spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled`, is introduced to enable/disable the feature. By default it is disabled, so the behavior should still be the same as before.

### How was this patch tested?

Added new tests in `KeyGroupedPartitioningSuite`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@sunchao do you plan to add this for 3.4? cc @xinrong-meng FYI

**Comment by sunchao** [comment]
@HyukjinKwon it'd be great if this can make it into 3.4, but it's not a release blocker.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1424,6 +1424,18 @@ object SQLConf {
...
 
+  val V2_BUCKETING_PARTIALLY_CLUSTERED_DISTRIBUTION_ENABLED =
+    buildConf("spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled")
+        .doc("During a storage-partitioned join, whether to allow input partitions to be " +
+            "partially clustered, when both sides of the join are of KeyGroupedPartitioning. At " +
+            "planning time, Spark will pick the side with less data size based on table " +
+            "statistics, group and replicate them to match the other side. This can help to " +
+            "reduce data skewness when certain partition values have large amount of data. This " +
+            s"config requires both ${V2_BUCKETING_ENABLED.key} and " +
+            s"${V2_BUCKETING_PUSH_PART_VALUES_ENABLED.key} to be enabled")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1435]: We need a version.
```scala
.version("3.4.0")
```

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (with only one minor comment about config version). Thank you so much, @sunchao .

cc @aokolnychyi , @viirya , @huaxingao , @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -1424,6 +1424,18 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val V2_BUCKETING_PARTIALLY_CLUSTERED_DISTRIBUTION_ENABLED =
+    buildConf("spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled")
+        .doc("During a storage-partitioned join, whether to allow input partitions to be " +
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1429]: Indentation?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala [hunk 3]
```diff
@@ -242,8 +246,16 @@ abstract class InMemoryBaseTable(
     data.foreach(_.rows.foreach { row =>
       val key = getKey(row, writeSchema)
       dataMap += dataMap.get(key)
-        .map(key -> _.withRow(row))
-        .getOrElse(key -> new BufferedRows(key).withRow(row))
+          .map { splits =>
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:249]: Indentation?

**Comment by cloud-fan** [comment]
Are we replicating the AQE skew join optimization in `EnsureRequirements`? I'm a bit worried about direction as `EnsureRequirements` is becoming more and more complicated.

**Comment by cloud-fan** [comment]
Since the storage partitioned join becomes more and more complicated, shall we revisit the framework? Essentially we need to join the partitions from two v2 tables at the driver side, and launch a task to process each partition pair. One idea is to add a physical rule before `EnsureRequirements`. This new rule matches join of two v2 tables, applies a join strategy, and sets a special flag in the join node so that `EnsureRequirements` can skip it. This is similar to `ShuffledJoin.isSkewJoin`:
```
trait ShuffledJoin extends JoinCodegenSupport {
  def isSkewJoin: Boolean

    override def requiredChildDistribution: Seq[Distribution] = {
    if (isSkewJoin) {
      // We re-arrange the shuffle partitions to deal with skew join, and the new children
      // partitioning doesn't satisfy `HashClusteredDistribution`.
      UnspecifiedDistribution :: UnspecifiedDistribution :: Nil
    } else {
      ClusteredDistribution(leftKeys) :: ClusteredDistribution(rightKeys) :: Nil
    }
  }
}
```

**Comment by sunchao** [comment]
@cloud-fan the idea is similar to skew join but for v2 sources, let me try to split the code into a separate rule following your idea.

**Comment by sunchao** [comment]
Hmm @cloud-fan , it seems a bit tricky to split the logic specific to SPJ out of `EnsureRequirements`, since it depends on `reorderJoinPredicates`, which is currently co-mingled with `EnsureRequirements`. I was trying to split `reorderJoinPredicates` into a separate rule but just realized I hit [SPARK-22042](https://issues.apache.org/jira/browse/SPARK-22042).

**Comment by dongjoon-hyun** [comment]
Shall we convert to `Ready for review` status back, @sunchao ?

**Comment by sunchao** [comment]
Yes, converted.

I found it is quite difficult to move the logic out of `EnsureRequirements`, because, as mentioned above, the optimization also depends on `reorderJoinPredicates` which needs to be executed together with `ensureDistributionAndOrdering` in a lock-step as we go up the query plan.

In addition, even if I extract out `reorderJoinPredicates` into a separate util method or something, it's still difficult to make the optimization as a separate rule because it also depends on certain part of the logic in `ensureDistributionAndOrdering` (for instance, as we go up the query plan tree, we'd expect join branches that have incompatible `KeyGroupedPartitioning`s to be handled by `ensureDistributionAndOrdering` and converted to use hash partitioning, so that we should not try to apply the optimization again on those branches).

As a compromise, I've extracted all special logic related to `KeyGroupedPartitioning` into a separate method `checkKeyGroupCompatible`, so the main body of `ensureDistributionAndOrdering` now looks much simpler. If necessary, I can also move this method (and related) into a separate class file instead, to make them more isolated.

**Comment by dongjoon-hyun** [comment]
Thank you, @sunchao !

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -1432,6 +1432,19 @@ object SQLConf {
...
+  val V2_BUCKETING_PARTIALLY_CLUSTERED_DISTRIBUTION_ENABLED =
+    buildConf("spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled")
+      .doc("During a storage-partitioned join, whether to allow input partitions to be " +
+        "partially clustered, when both sides of the join are of KeyGroupedPartitioning. At " +
+        "planning time, Spark will pick the side with less data size based on table " +
+        "statistics, group and replicate them to match the other side. This is an optimization " +
+        "on skew join and can help to reduce data skewness when certain partitions are assigned " +
+        s"large amount of data. This config requires both ${V2_BUCKETING_ENABLED.key} and " +
+        s"${V2_BUCKETING_PUSH_PART_VALUES_ENABLED.key} to be enabled")
+      .version("3.4.0")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1444]: If you don't mind, we still want to have this in 3.4.0, @cloud-fan and @xinrong-meng . Please let us know your opinions.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 5]
```diff
@@ -40,15 +41,18 @@ case class BatchScanExec(
...
 
   @transient lazy val batch = scan.toBatch
 
   // TODO: unify the equal/hashCode implementation for all data source v2 query plans.
   override def equals(other: Any): Boolean = other match {
     case other: BatchScanExec =>
       this.batch == other.batch && this.runtimeFilters == other.runtimeFilters &&
-          this.commonPartitionValues == other.commonPartitionValues
+          this.commonPartitionValues == other.commonPartitionValues &&
+          this.replicatePartitions == other.replicatePartitions
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:55]: shall we also check `applyPartialClustering`?

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 6]
```diff
@@ -411,6 +336,235 @@ case class EnsureRequirements(
...
+          left, mergedPartValues, applyPartialClustering, replicateLeftSide)
+        newRight = populatePartitionValues(
+          right, mergedPartValues, applyPartialClustering, replicateRightSide)
+      }
+    }
+
+    if (isCompatible) Some(Seq(newLeft, newRight)) else None
+  }
+
+  // Similar to `OptimizeSkewedJoin.canSplitRightSide`
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:503]: This comment looks not informative.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 7]
```diff
@@ -411,6 +336,235 @@ case class EnsureRequirements(
...
+        // that is, multiple partitions with the same partition value. In the
+        // following, we calculate how many partitions that each distinct partition
+        // value has, and pushdown the information to scans, so they can adjust their
+        // final input partitions respectively.
+        if (conf.v2BucketingPartiallyClusteredDistributionEnabled) {
+          logInfo("Calculating partially clustered distribution for " +
+              "storage-partitioned join")
+
+          val canReplicateLeft = canReplicateLeftSide(joinType)
+          val canReplicateRight = canReplicateRightSide(joinType)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:435]: Might be good to add a comment for disallowed cases (specified sides in outer join) and the reason.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 8]
```diff
@@ -132,12 +140,66 @@ case class BatchScanExec(
...
       outputPartitioning match {
         case p: KeyGroupedPartitioning =>
-          val partitionMapping = finalPartitions.map(s =>
-            s.head.asInstanceOf[HasPartitionKey].partitionKey() -> s).toMap
-          finalPartitions = p.partitionValues.map { partValue =>
-            // Use empty partition for those partition values that are not present
-            partitionMapping.getOrElse(partValue, Seq.empty)
+          // This means the input partitions are not grouped by partition values. We'll need to
+          // check `groupByPartitionValues` and decide whether to group and replicate splits within
+          // a partition.
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:145]: Does this comment need to be moved below on top of `if (commonPartitionKeys.isDefined ...`?

**Review by viirya — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Could you resolve the conflicts once more, @sunchao ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala (hunk 8 above)
Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala]: you're right - moved.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala (hunk 7 above)
Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:439]: +1. Added comments for this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala (hunk 6 above)
Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:507]: I wanted to point out these two methods are identical to the ones in `OptimizeSkewedJoin` except the method name

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala (hunk 5 above)
Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala]: Oops good catch

**Review by dongjoon-hyun — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Merged to master/3.4.

Thank you so much, @sunchao , @cloud-fan , @viirya .

Also, cc @aokolnychyi and @RussellSpitzer

**Comment by sunchao** [comment]
Thanks @cloud-fan @dongjoon-hyun @viirya !

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 9]
```diff
@@ -39,15 +40,19 @@ case class BatchScanExec(
     keyGroupedPartitioning: Option[Seq[Expression]] = None,
     ordering: Option[Seq[SortOrder]] = None,
     @transient table: Table,
-    commonPartitionValues: Option[Seq[InternalRow]] = None) extends DataSourceV2ScanExecBase {
+    commonPartitionValues: Option[Seq[(InternalRow, Int)]] = None,
+    applyPartialClustering: Boolean = false,
+    replicatePartitions: Boolean = false) extends DataSourceV2ScanExecBase {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:45]: This happens quite a few times that we add a new parameter to `BatchScanExec` and change a lot of code which does a pattern matching over `BatchScanExec`. Can we put SPJ-related information in a wrapper class? Then `BatchScanExec` can be more stable. cc @sunchao

Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:45]: @cloud-fan yes, it's a good idea. The issue is quite annoying. Let me create a PR for it.

**Comment by anigos** [comment]
Partially Clustered Distribution doesn't work with Full Outer Joins for example when we are doing merge. In this cases it produces redundant dataset. Is there any development going on for that? @sunchao @viirya @dongjoon-hyun

**Pull request closed without merging**